#building context based recommendation syystem

In [ ]:
import zipfile
import os

zip_file_name = '/archive.zip'
if os.path.exists(zip_file_name):
    with zipfile.ZipFile(zip_file_name, 'r') as zip_ref:
        zip_ref.extractall('.')
    print(f"Extracted {zip_file_name}")
else:
    print(f"{zip_file_name} not found.")

Extracted /archive.zip


In [ ]:
import pandas as pd
try:
    movies_df = pd.read_csv('tmdb_5000_movies.csv')
    credits_df = pd.read_csv('tmdb_5000_credits.csv')

    print("Movies DataFrame:")
    display(movies_df.sample())

    print("\nCredits DataFrame:")
    display(credits_df.head())

except FileNotFoundError:
    print("Make sure 'tmdb_5000_movies.csv' and 'tmdb_5000_credits.csv' are in the current directory.")
except Exception as e:
    print(f"An error occurred: {e}")

Movies DataFrame:


budget                                             genres homepage  \
990  50000000  [{"id": 18, "name": "Drama"}, {"id": 36, "name...      NaN   

        id                                           keywords  \
990  10533  [{"id": 818, "name": "based on novel"}, {"id":...   

    original_language      original_title  \
990                en  The Scarlet Letter   

                                              overview  popularity  \
990  Set in puritanical Boston in the mid 1600s, th...     7.44268   

                                  production_companies  \
990  [{"name": "Hollywood Pictures", "id": 915}, {"...   

                                  production_countries release_date   revenue  \
990  [{"iso_3166_1": "US", "name": "United States o...   1995-10-13  10382407   

     runtime                          spoken_languages    status  \
990    135.0  [{"iso_639_1": "en", "name": "English"}]  Released   

                                               tagline               title  \
990  When intimacy is forbidden and passion is a si...  The Scarlet Letter   

     vote_average  vote_count  
990           5.5         108


Credits DataFrame:


movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [ ]:
#credits_df.sample(2)['crew'].values
credits_df.sample(2)['cast'].values

array(['[{"cast_id": 1, "character": "Jackson Pollock", "credit_id": "52fe44f19251416c75046421", "gender": 2, "id": 228, "name": "Ed Harris", "order": 0}, {"cast_id": 2, "character": "Lee Krasner", "credit_id": "52fe44f19251416c75046425", "gender": 1, "id": 4726, "name": "Marcia Gay Harden", "order": 1}, {"cast_id": 3, "character": "Dan Miller", "credit_id": "52fe44f19251416c75046429", "gender": 2, "id": 19453, "name": "Tom Bower", "order": 2}, {"cast_id": 4, "character": "Ruth Kligman", "credit_id": "52fe44f19251416c7504642d", "gender": 1, "id": 6161, "name": "Jennifer Connelly", "order": 3}, {"cast_id": 5, "character": "Howard Putzel", "credit_id": "52fe44f19251416c75046431", "gender": 2, "id": 4971, "name": "Bud Cort", "order": 4}, {"cast_id": 6, "character": "Tony Smith", "credit_id": "52fe44f19251416c75046435", "gender": 2, "id": 11512, "name": "John Heard", "order": 5}, {"cast_id": 7, "character": "Willem DeKooning", "credit_id": "52fe44f19251416c75046439", "gender": 2, "id": 557

In [ ]:
movies = movies_df.merge(credits_df , on ="title")

In [ ]:
movies.sample(1)

budget                          genres homepage      id  \
1072  40000000  [{"id": 35, "name": "Comedy"}]      NaN  232672   

                                               keywords original_language  \
1072  [{"id": 409, "name": "africa"}, {"id": 4513, "...                en   

     original_title                                           overview  \
1072        Blended  After a bad blind date, a man and woman find t...   

      popularity                               production_companies  ...  \
1072   44.120328  [{"name": "Happy Madison Productions", "id": 2...  ...   

     runtime                          spoken_languages    status  \
1072   117.0  [{"iso_639_1": "en", "name": "English"}]  Released   

                                            tagline    title vote_average  \
1072  Single Dad, No Clue. Single Mum, Flying Solo.  Blended          6.6   

     vote_count movie_id                                               cast  \
1072       1006   232672  [{"cast_id": 2, "character": "Lauren", "credit...   

                                                   crew  
1072  [{"credit_id": "53e00e910e0a265a99005231", "de...  

[1 rows x 23 columns]

In [ ]:
movies.info(),movies.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4809 entries, 0 to 4808
Data columns (total 23 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4809 non-null   int64  
 1   genres                4809 non-null   object 
 2   homepage              1713 non-null   object 
 3   id                    4809 non-null   int64  
 4   keywords              4809 non-null   object 
 5   original_language     4809 non-null   object 
 6   original_title        4809 non-null   object 
 7   overview              4806 non-null   object 
 8   popularity            4809 non-null   float64
 9   production_companies  4809 non-null   object 
 10  production_countries  4809 non-null   object 
 11  release_date          4808 non-null   object 
 12  revenue               4809 non-null   int64  
 13  runtime               4807 non-null   float64
 14  spoken_languages      4809 non-null   object 
 15  status               

(None,
              budget             id   popularity       revenue      runtime  \
 count  4.809000e+03    4809.000000  4809.000000  4.809000e+03  4807.000000   
 mean   2.902780e+07   57120.571429    21.491664  8.227511e+07   106.882255   
 std    4.070473e+07   88653.369849    31.803366  1.628379e+08    22.602535   
 min    0.000000e+00       5.000000     0.000000  0.000000e+00     0.000000   
 25%    7.800000e+05    9012.000000     4.667230  0.000000e+00    94.000000   
 50%    1.500000e+07   14624.000000    12.921594  1.917000e+07   103.000000   
 75%    4.000000e+07   58595.000000    28.350529  9.291317e+07   118.000000   
 max    3.800000e+08  459488.000000   875.581305  2.787965e+09   338.000000   
 
        vote_average    vote_count       movie_id  
 count   4809.000000   4809.000000    4809.000000  
 mean       6.092514    690.331670   57120.571429  
 std        1.193989   1234.187111   88653.369849  
 min        0.000000      0.000000       5.000000  
 25%        5.600000

#slecting col which used to generate tags

In [ ]:
"""
#skiiping num col

genres
id
keywords
title
overview
cast
crew
"""

movies = movies[['id','title','overview','keywords','genres','cast','crew']]

In [ ]:
movies.sample()

id             title  \
2121  15655  The Tigger Movie   

                                               overview  \
2121  As it happens, everybody - Pooh, Piglet, Eeyor...   

                                               keywords  \
2121  [{"id": 3905, "name": "owl"}, {"id": 4144, "na...   

                                                 genres  \
2121  [{"id": 16, "name": "Animation"}, {"id": 10751...   

                                                   cast  \
2121  [{"cast_id": 2, "character": "Tigger / Winnie ...   

                                                   crew  
2121  [{"credit_id": "59121da0c3a3686519043247", "de...

In [ ]:
#creating new data feame from movies

In [ ]:
#and creating tags col from this

In [ ]:
#correcting formaat off col

In [ ]:
movies.isnull().sum()

id          0
title       0
overview    3
keywords    0
genres      0
cast        0
crew        0
dtype: int64

In [ ]:
movies.dropna(inplace=True)

In [ ]:
movies.duplicated().sum()

np.int64(0)

In [ ]:
movies.iloc[0].genres

'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]

covert into ['action,advebture,...]

In [ ]:
#convert dict key into number expecting no

In [ ]:
import ast
ast.literal_eval('[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]')


[{'id': 28, 'name': 'Action'},
 {'id': 12, 'name': 'Adventure'},
 {'id': 14, 'name': 'Fantasy'},
 {'id': 878, 'name': 'Science Fiction'}]

In [ ]:
def convert(text):
    L = []
    for i in ast.literal_eval(text):
        L.append(i['name'])
    return L

In [ ]:
movies['genres'].apply(convert) #list of genres for allmoive

0       [Action, Adventure, Fantasy, Science Fiction]
1                        [Adventure, Fantasy, Action]
2                          [Action, Adventure, Crime]
3                    [Action, Crime, Drama, Thriller]
4                [Action, Adventure, Science Fiction]
                            ...                      
4804                        [Action, Crime, Thriller]
4805                                [Comedy, Romance]
4806               [Comedy, Drama, Romance, TV Movie]
4807                                               []
4808                                    [Documentary]
Name: genres, Length: 4806, dtype: object

In [ ]:
movies['genres'] = movies['genres'].apply(convert)

In [ ]:
movies.sample()

id                     title  \
3194  27329  The Legend of Suriyothai   

                                               overview  \
3194  During the 16th century, as Thailand contends ...   

                                               keywords  \
3194  [{"id": 2011, "name": "queen"}, {"id": 3434, "...   

                              genres  \
3194  [Drama, Foreign, History, War]   

                                                   cast  \
3194  [{"cast_id": 1001, "character": "Queen Suriyot...   

                                                   crew  
3194  [{"credit_id": "579b13ab9251414cd9006566", "de...

In [ ]:
#applying to rest col

In [ ]:
movies['keywords'] = movies['keywords'].apply(convert)

In [ ]:
display(movies.iloc[670])

id                                                      12618
title                                           Random Hearts
overview    After the death of their loved ones in a tragi...
keywords    [infidelity, politician, airplane crash, death...
genres                                       [Drama, Romance]
cast        [{"cast_id": 1, "character": "Dutch Van Den Br...
crew        [{"credit_id": "52fe45089251416c75049655", "de...
Name: 670, dtype: object

cast name should be orignal name
like when think is this moive of tom cruise i gonna watch top 3 only

In [ ]:
def convert2(obj):
    L = []
    counter =0
    for i in ast.literal_eval(obj):
      if counter !=3:
         L.append(i['name'])
         counter+=1
      else:
        break
    return L

In [ ]:
movies['cast'] = movies['cast'].apply(convert2)

In [ ]:
movies.head(3)

id                                     title  \
0   19995                                    Avatar   
1     285  Pirates of the Caribbean: At World's End   
2  206647                                   Spectre   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   

                                               cast  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]   
1     [Johnny Depp, Orlando Bloom, Keira Knightley]   
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...

In [ ]:
#for crew

In [ ]:
movies['crew'][0]

'[{"credit_id": "52fe48009251416c750aca23", "department": "Editing", "gender": 0, "id": 1721, "job": "Editor", "name": "Stephen E. Rivkin"}, {"credit_id": "539c47ecc3a36810e3001f87", "department": "Art", "gender": 2, "id": 496, "job": "Production Design", "name": "Rick Carter"}, {"credit_id": "54491c89c3a3680fb4001cf7", "department": "Sound", "gender": 0, "id": 900, "job": "Sound Designer", "name": "Christopher Boyes"}, {"credit_id": "54491cb70e0a267480001bd0", "department": "Sound", "gender": 0, "id": 900, "job": "Supervising Sound Editor", "name": "Christopher Boyes"}, {"credit_id": "539c4a4cc3a36810c9002101", "department": "Production", "gender": 1, "id": 1262, "job": "Casting", "name": "Mali Finn"}, {"credit_id": "5544ee3b925141499f0008fc", "department": "Sound", "gender": 2, "id": 1729, "job": "Original Music Composer", "name": "James Horner"}, {"credit_id": "52fe48009251416c750ac9c3", "department": "Directing", "gender": 2, "id": 2710, "job": "Director", "name": "James Cameron"},

In [ ]:
#need name of director only [every moive have one director]

In [ ]:
def fetch_director(text):
    L = []
    for i in ast.literal_eval(text):
        if i['job'] == 'Director':
            L.append(i['name'])
    return L

In [ ]:
movies['crew'] = movies['crew'].apply(fetch_director)

In [ ]:
display(movies.iloc[670])

id                                                      12618
title                                           Random Hearts
overview    After the death of their loved ones in a tragi...
keywords    [infidelity, politician, airplane crash, death...
genres                                       [Drama, Romance]
cast        [Harrison Ford, Kristin Scott Thomas, Charles ...
crew                                         [Sydney Pollack]
Name: 670, dtype: object

In [ ]:
movies['overview'][0] #string

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.'

In [ ]:
#convert into list to concat with other coulumns
movies['overview'].apply(lambda x:x.split())

0       [In, the, 22nd, century,, a, paraplegic, Marin...
1       [Captain, Barbossa,, long, believed, to, be, d...
2       [A, cryptic, message, from, Bond’s, past, send...
3       [Following, the, death, of, District, Attorney...
4       [John, Carter, is, a, war-weary,, former, mili...
                              ...                        
4804    [El, Mariachi, just, wants, to, play, his, gui...
4805    [A, newlywed, couple's, honeymoon, is, upended...
4806    ["Signed,, Sealed,, Delivered", introduces, a,...
4807    [When, ambitious, New, York, attorney, Sam, is...
4808    [Ever, since, the, second, grade, when, he, fi...
Name: overview, Length: 4806, dtype: object

In [ ]:
movies['overview'] = movies['overview'].apply(lambda x:x.split())

In [ ]:
movies.tail(3)

id                      title  \
4806  231617  Signed, Sealed, Delivered   
4807  126186           Shanghai Calling   
4808   25975          My Date with Drew   

                                               overview  \
4806  ["Signed,, Sealed,, Delivered", introduces, a,...   
4807  [When, ambitious, New, York, attorney, Sam, is...   
4808  [Ever, since, the, second, grade, when, he, fi...   

                                               keywords  \
4806  [date, love at first sight, narration, investi...   
4807                                                 []   
4808          [obsession, camcorder, crush, dream girl]   

                                  genres  \
4806  [Comedy, Drama, Romance, TV Movie]   
4807                                  []   
4808                       [Documentary]   

                                                   cast  \
4806         [Eric Mabius, Kristin Booth, Crystal Lowe]   
4807          [Daniel Henney, Eliza Coupe, Bill Paxton]   
4808  [Drew Barrymore, Brian Herzlinger, Corey Feldman]   

                                          crew  
4806                             [Scott Smith]  
4807                             [Daniel Hsia]  
4808  [Brian Herzlinger, Jon Gunn, Brett Winn]

In [ ]:
#concating all into list the string

keywords	genres	cast	crew applying tranformation to remove spaces to avoide same name confusion

'Daniel Hsia' have 2 tag model get confuse
'DanielHsia'  have 1 tag 1 entity

In [ ]:
movies['genres'].apply(lambda x:[i.replace(" ","") for i in x])

0       [Action, Adventure, Fantasy, ScienceFiction]
1                       [Adventure, Fantasy, Action]
2                         [Action, Adventure, Crime]
3                   [Action, Crime, Drama, Thriller]
4                [Action, Adventure, ScienceFiction]
                            ...                     
4804                       [Action, Crime, Thriller]
4805                               [Comedy, Romance]
4806               [Comedy, Drama, Romance, TVMovie]
4807                                              []
4808                                   [Documentary]
Name: genres, Length: 4806, dtype: object

In [ ]:
#apply to all 3
def collapse(L):
    L1 = []
    for i in L:
        L1.append(i.replace(" ",""))
    return L1

In [ ]:
movies['cast'] = movies['cast'].apply(collapse)
movies['crew'] = movies['crew'].apply(collapse)
movies['genres'] = movies['genres'].apply(collapse)
movies['keywords'] = movies['keywords'].apply(collapse)

In [ ]:
movies.head(4)

id                                     title  \
0   19995                                    Avatar   
1     285  Pirates of the Caribbean: At World's End   
2  206647                                   Spectre   
3   49026                     The Dark Knight Rises   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   
1  [ocean, drugabuse, exoticisland, eastindiatrad...   
2  [spy, basedonnovel, secretagent, sequel, mi6, ...   
3  [dccomics, crimefighter, terrorist, secretiden...   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   
1                  [Adventure, Fantasy, Action]   
2                    [Action, Adventure, Crime]   
3              [Action, Crime, Drama, Thriller]   

                                            cast                crew  
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]      [JamesCameron]  
1     [JohnnyDepp, OrlandoBloom, KeiraKnightley]     [GoreVerbinski]  
2      [DanielCraig, ChristophWaltz, LéaSeydoux]         [SamMendes]  
3      [ChristianBale, MichaelCaine, GaryOldman]  [ChristopherNolan]

In [ ]:
#concating

In [ ]:
movies['tags'] = movies['overview']+  movies['genres']+ movies['keywords']+movies['cast']+ movies['crew']

In [ ]:
movies.head(1)

id   title                                           overview  \
0  19995  Avatar  [In, the, 22nd, century,, a, paraplegic, Marin...   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   

                                            cast            crew  \
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]  [JamesCameron]   

                                                tags  
0  [In, the, 22nd, century,, a, paraplegic, Marin...

In [ ]:
#removing these 5 no need beacuse all info have 1 col

In [ ]:
new_df = movies[["id","title","tags"]]

In [ ]:
new_df.sample(2)

id                title  \
379   9387  Conan the Barbarian   
1610   865      The Running Man   

                                                   tags  
379   [A, film, adaptation, of, the, classic, sword,...  
1610  [By, 2017,, the, global, economy, has, collaps...

In [ ]:
#list to string for tags

In [ ]:
new_df["tags"].apply(lambda x: " ".join(x)) #corpus

0       In the 22nd century, a paraplegic Marine is di...
1       Captain Barbossa, long believed to be dead, ha...
2       A cryptic message from Bond’s past sends him o...
3       Following the death of District Attorney Harve...
4       John Carter is a war-weary, former military ca...
                              ...                        
4804    El Mariachi just wants to play his guitar and ...
4805    A newlywed couple's honeymoon is upended by th...
4806    "Signed, Sealed, Delivered" introduces a dedic...
4807    When ambitious New York attorney Sam is sent t...
4808    Ever since the second grade when he first saw ...
Name: tags, Length: 4806, dtype: object

In [ ]:
new_df["tags"] = new_df["tags"].apply(lambda x: " ".join(x))

/tmp/ipython-input-3719869087.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df["tags"] = new_df["tags"].apply(lambda x: " ".join(x))


In [ ]:
new_df["tags"].sample(2),new_df["tags"][2]

(2888    The late 1940s. Richard Langley, a bachelor pl...
 146     Alex, Marty, Gloria and Melman are still tryin...
 Name: tags, dtype: object,
 'A cryptic message from Bond’s past sends him on a trail to uncover a sinister organization. While M battles political forces to keep the secret service alive, Bond peels back the layers of deceit to reveal the terrible truth behind SPECTRE. Action Adventure Crime spy basedonnovel secretagent sequel mi6 britishsecretservice unitedkingdom DanielCraig ChristophWaltz LéaSeydoux SamMendes')

In [ ]:
new_df["tags"].apply(lambda x:x.lower()) #converting into lower case

0       in the 22nd century, a paraplegic marine is di...
1       captain barbossa, long believed to be dead, ha...
2       a cryptic message from bond’s past sends him o...
3       following the death of district attorney harve...
4       john carter is a war-weary, former military ca...
                              ...                        
4804    el mariachi just wants to play his guitar and ...
4805    a newlywed couple's honeymoon is upended by th...
4806    "signed, sealed, delivered" introduces a dedic...
4807    when ambitious new york attorney sam is sent t...
4808    ever since the second grade when he first saw ...
Name: tags, Length: 4806, dtype: object

In [ ]:
new_df["tags"] = new_df["tags"].apply(lambda x:x.lower())

/tmp/ipython-input-3481706753.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df["tags"] = new_df["tags"].apply(lambda x:x.lower())


In [ ]:
#replacing similar word
#applying steaming to replace by one similar word

In [ ]:
import nltk
from nltk.stem.porter import PorterStemmer
ps = PorterStemmer()

In [ ]:
ps.stem("wording")

'word'

In [ ]:
def stem(text):
  y = []

  for i in text.split():
    y.append(ps.stem(i)) #into list the string again

  return " ".join(y)


In [ ]:
new_df["tags"][0]

'in the 22nd century, a paraplegic marine is dispatched to the moon pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization. action adventure fantasy sciencefiction cultureclash future spacewar spacecolony society spacetravel futuristic romance space alien tribe alienplanet cgi marine soldier battle loveaffair antiwar powerrelations mindandsoul 3d samworthington zoesaldana sigourneyweaver jamescameron'

In [ ]:
stem("in the 22nd century, a paraplegic marine is dispatched to the moon pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization. action adventure fantasy sciencefiction cultureclash future spacewar spacecolony society spacetravel futuristic romance space alien tribe alienplanet cgi marine soldier battle loveaffair antiwar powerrelations mindandsoul 3d samworthington zoesaldana sigourneyweaver jamescameron")

'in the 22nd century, a parapleg marin is dispatch to the moon pandora on a uniqu mission, but becom torn between follow order and protect an alien civilization. action adventur fantasi sciencefict cultureclash futur spacewar spacecoloni societi spacetravel futurist romanc space alien tribe alienplanet cgi marin soldier battl loveaffair antiwar powerrel mindandsoul 3d samworthington zoesaldana sigourneyweav jamescameron'

In [ ]:
new_df["tags"].apply(stem)

0       in the 22nd century, a parapleg marin is dispa...
1       captain barbossa, long believ to be dead, ha c...
2       a cryptic messag from bond’ past send him on a...
3       follow the death of district attorney harvey d...
4       john carter is a war-weary, former militari ca...
                              ...                        
4804    el mariachi just want to play hi guitar and ca...
4805    a newlyw couple' honeymoon is upend by the arr...
4806    "signed, sealed, delivered" introduc a dedic q...
4807    when ambiti new york attorney sam is sent to s...
4808    ever sinc the second grade when he first saw h...
Name: tags, Length: 4806, dtype: object

In [ ]:
new_df["tags"] = new_df["tags"].apply(stem)

/tmp/ipython-input-825579503.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df["tags"] = new_df["tags"].apply(stem)


In [ ]:
new_df.sample(2)

id                             title  \
1854   746                  The Last Emperor   
4058  9731  Friday the 13th: A New Beginning   

                                                   tags  
1854  a dramat histori of pu yi, the last of the emp...  
4058  homicid maniac jason return from the grave to ...

In [ ]:
new_df.isnull().sum()

id       0
title    0
tags     0
dtype: int64

In [ ]:
new_df.duplicated().sum()

np.int64(0)

In [ ]:
#how to know which moive should recommend
#to find similarity score between text
#vectorization to cal similarity score

name of moive becoome vector  to finding vector in NDim space

'''BOW'''

In [ ]:
#bow all word convert into vector
#by calculating freq of word from vector get extract for every moive tags
#convert the vector table most common freq (1,2,3,4,5,8)
#then finding clost vector of other moive drown by bow

In [ ]:
import sklearn
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features=7000,stop_words='english') #stop_word removel

In [ ]:
vectors = cv.fit_transform(new_df["tags"]).toarray()

In [ ]:
vectors[2]

array([0, 0, 0, ..., 0, 0, 0])

In [ ]:
len(cv.get_feature_names_out()),cv.get_feature_names_out()

(7000,
 array(['000', '007', '10', ..., 'zoo', 'zooey', 'zooeydeschanel'],
       dtype=object))

In [ ]:
#no of moives == no of vector of 7000 dim
#cal distance for moive cosine (angle btween them)
#which ever vector is have less angle then we can say thhese moive can recommnede

In [ ]:
#less simlilarity more distance/angle
#more similarity then less ditance between vector in dim/angle og vector drawn

Cosine Similarity

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity

In [ ]:
similarity = cosine_similarity(vectors)

In [ ]:
cosine_similarity(vectors).shape #every moive cal distance from each one

(4806, 4806)

In [ ]:
#4806*4806 disance cal

In [ ]:
similarity[0],similarity[0].shape,similarity[1]

(array([1.        , 0.07537784, 0.07756315, ..., 0.04141577, 0.        ,
        0.        ]),
 (4806,),
 array([0.07537784, 1.        , 0.0571662 , ..., 0.02289343, 0.        ,
        0.02512595]))

In [ ]:
# similarity of first moive with every moive this vector
#extract evey moive index from array (and rest is distance from other moive) and
#and sort distance (distance have less is get lcose and other will get back top 5 will recommend)

In [ ]:
new_df[new_df["title"] == 'Batman Begins'].index[0] #fetch index

np.int64(119)

In [223]:
sorted(list(enumerate(similarity[0])),reverse = True,key=lambda x:x[1])[1:6]
#due to sort all thing sort moive value using enuemrate fixing first and cal
#simlar score and see top 5

[(1214, np.float64(0.2528249484237158)),
 (2405, np.float64(0.25087260300212727)),
 (3728, np.float64(0.23596995186213474)),
 (539, np.float64(0.23268946049775863)),
 (507, np.float64(0.22975187432024505))]

In [224]:
def recommend(moive):
  movie_index =new_df[new_df["title"] == moive].index[0] #get index
  distances = similarity[movie_index] #find similarity
  movies_list = sorted(list(enumerate(distances)),reverse = True,key=lambda x:x[1])[1:6]

  for i in movies_list:
    print(i[0])

In [225]:
recommend("Avatar")

1214
2405
3728
539
507


In [232]:
new_df.iloc[0].title,new_df.iloc[1214].title,new_df.iloc[2405].title,new_df.iloc[3728].title,new_df.iloc[539].title,new_df.iloc[507].title

('Avatar',
 'Aliens vs Predator: Requiem',
 'Aliens',
 'Falcon Rising',
 'Titan A.E.',
 'Independence Day')

In [233]:
get_recommended_movies("Avatar"),get_recommended_movies("Batman Begins")

Recommendations for 'Avatar':
Aliens vs Predator: Requiem
Aliens
Falcon Rising
Titan A.E.
Independence Day
Recommendations for 'Batman Begins':
The Dark Knight
Batman
Batman
The Dark Knight Rises
10th & Wolf


(None, None)

In [243]:
import pickle
pickle.dump(new_df,open('movies.pkl','wb'))
pickle.dump(similarity,open('similarity.pkl','wb'))

instead of sending full datafram sending dict

In [245]:
pickle.dump(new_df.to_dict(),open('movies_dict.pkl','wb'))

In [ ]:
# to genrate pip freeze > requirements.txt
#web: sh setup.sh && streamlit run app.py  procfile
# venv .gitignore
"""mkdir -p ~/.streamlit/

echo "\
[server]\n\
port = $PORT\n\
enableCORS = fasle\n\
headless = true\n\
\n\
" > ~/.streamlit/config.toml""" setup.sh
#

# Task
Evaluate the recommendation engine code using appropriate metrics.

## Prepare evaluation data

### Subtask:
Split the dataset into training and testing sets. The training set will be used to build the recommendation engine, and the testing set will be used to evaluate its performance.


**Reasoning**:
Split the data into training and testing sets to prepare for model evaluation.



In [234]:
from sklearn.model_selection import train_test_split

new_df_train, new_df_test = train_test_split(new_df, test_size=0.2, random_state=42)

## Implement evaluation metrics

### Subtask:
Write functions to calculate relevant evaluation metrics such as Precision, Recall, or NDCG.


**Reasoning**:
Define functions to calculate Precision, Recall, and NDCG for evaluating the recommendation engine.



In [235]:
def calculate_precision(actual_relevant, recommended_items):
    """Calculates Precision@k.

    Args:
        actual_relevant: A set of actual relevant item IDs.
        recommended_items: A list of recommended item IDs.

    Returns:
        The precision score.
    """
    if not recommended_items:
        return 0.0
    relevant_recommended = len(set(recommended_items) & set(actual_relevant))
    return relevant_recommended / len(recommended_items)

def calculate_recall(actual_relevant, recommended_items):
    """Calculates Recall@k.

    Args:
        actual_relevant: A set of actual relevant item IDs.
        recommended_items: A list of recommended item IDs.

    Returns:
        The recall score.
    """
    if not actual_relevant:
        return 0.0
    relevant_recommended = len(set(recommended_items) & set(actual_relevant))
    return relevant_recommended / len(actual_relevant)

def calculate_ndcg(actual_relevant, recommended_items, relevance_scores=None):
    """Calculates NDCG@k.

    Args:
        actual_relevant: A set of actual relevant item IDs.
        recommended_items: A list of recommended item IDs.
        relevance_scores: A dictionary mapping item IDs to their relevance scores.
                          If None, relevance is assumed to be 1 for relevant items
                          and 0 for irrelevant items.

    Returns:
        The NDCG score.
    """
    if not recommended_items:
        return 0.0

    dcg = 0.0
    idcg = 0.0

    # Calculate DCG
    for i, item in enumerate(recommended_items):
        rank = i + 1
        if relevance_scores:
            relevance = relevance_scores.get(item, 0)
        else:
            relevance = 1 if item in actual_relevant else 0
        dcg += (2**relevance - 1) / (rank + 1) if rank > 1 else (2**relevance - 1)

    # Calculate IDCG
    ideal_list = sorted(list(actual_relevant), key=lambda x: relevance_scores.get(x, 1) if relevance_scores else 1, reverse=True)
    for i, item in enumerate(ideal_list[:len(recommended_items)]):
        rank = i + 1
        if relevance_scores:
            relevance = relevance_scores.get(item, 0)
        else:
            relevance = 1
        idcg += (2**relevance - 1) / (rank + 1) if rank > 1 else (2**relevance - 1)

    return dcg / idcg if idcg > 0 else 0.0

## Evaluate the model

### Subtask:
Use the implemented metrics to evaluate the performance of your recommendation engine on the testing set.


**Reasoning**:
Define the `get_recommended_movies` function and then iterate through the test set, get recommendations, determine actual relevant items, and calculate the evaluation metrics for each test movie.



**Reasoning**:
The error indicates an index out of bounds issue when accessing the `similarity` matrix using `test_movie_index`. This is likely because the `test_movie_index` refers to the original index in `new_df` before the train-test split, and this index might be larger than the size of the `similarity` matrix which was calculated based on the whole `new_df`. To fix this, the `similarity` calculation should be done based on the training data, and then the similarity for test movies should be calculated against the training data's vectors.



In [237]:
# Recalculate vectors and similarity based on the training set
cv = CountVectorizer(max_features=7000, stop_words='english')
vectors_train = cv.fit_transform(new_df_train["tags"]).toarray()
vectors_test = cv.transform(new_df_test["tags"]).toarray()

# Calculate similarity between test set vectors and training set vectors
similarity_test_train = cosine_similarity(vectors_test, vectors_train)

def get_recommended_movies_corrected(movie_title, k=5):
    """Gets top k recommended movie titles based on similarity to training data.

    Args:
        movie_title: The title of the movie in the test set.
        k: The number of recommendations to return.

    Returns:
        A list of recommended movie titles from the training set.
    """
    try:
        # Find the index of the movie in the test set
        test_movie_index_in_test_df = new_df_test[new_df_test["title"] == movie_title].index[0]
        # Get the corresponding row in the similarity_test_train matrix
        distances_to_train = similarity_test_train[new_df_test.index.get_loc(test_movie_index_in_test_df)]

        # Get the indices of the top k most similar movies in the training set
        movies_list = sorted(list(enumerate(distances_to_train)), reverse=True, key=lambda x: x[1])[:k] # Get top k, including the movie itself if present

        # Filter out the movie itself if it's in the training set and among top k
        recommended_movies = []
        for i, score in movies_list:
             recommended_movie_title = new_df_train.iloc[i].title
             if recommended_movie_title != movie_title:
                 recommended_movies.append(recommended_movie_title)
             if len(recommended_movies) == k:
                 break

        return recommended_movies
    except IndexError:
        print(f"Movie '{movie_title}' not found in the test dataset.")
        return []

# Determine a suitable value for k
k_recommendations = 10
similarity_threshold = 0.2 # Define a similarity threshold for actual relevant items based on training data

all_precision_scores = []
all_recall_scores = []
all_ndcg_scores = []

# Iterate through each movie in the test set
for index, row in new_df_test.iterrows():
    test_movie_title = row['title']

    # Get recommended items from the training set
    recommended_items = get_recommended_movies_corrected(test_movie_title, k=k_recommendations)

    # Determine actual relevant items from the training set based on similarity threshold
    # Calculate similarity of the test movie vector with all training movie vectors
    test_movie_index_in_test_df = new_df_test[new_df_test["title"] == test_movie_title].index[0]
    distances_to_train_for_relevant = similarity_test_train[new_df_test.index.get_loc(test_movie_index_in_test_df)]

    relevant_indices_in_train = [i for i, score in enumerate(distances_to_train_for_relevant) if score >= similarity_threshold]
    actual_relevant_items = set([new_df_train.iloc[i].title for i in relevant_indices_in_train])


    # Calculate metrics
    precision = calculate_precision(actual_relevant_items, recommended_items)
    recall = calculate_recall(actual_relevant_items, recommended_items)
    ndcg = calculate_ndcg(actual_relevant_items, recommended_items)

    all_precision_scores.append(precision)
    all_recall_scores.append(recall)
    all_ndcg_scores.append(ndcg)

# Calculate average metrics
average_precision = sum(all_precision_scores) / len(all_precision_scores) if all_precision_scores else 0
average_recall = sum(all_recall_scores) / len(all_recall_scores) if all_recall_scores else 0
average_ndcg = sum(all_ndcg_scores) / len(all_ndcg_scores) if all_ndcg_scores else 0

print(f"Average Precision@{k_recommendations}: {average_precision}")
print(f"Average Recall@{k_recommendations}: {average_recall}")
print(f"Average NDCG@{k_recommendations}: {average_ndcg}")

Average Precision@10: 0.7507276507276508
Average Recall@10: 0.5877176025182262
Average NDCG@10: 0.9646569646569647


## Analyze results

### Subtask:
Interpret the evaluation results to understand the strengths and weaknesses of your recommendation engine.


**Reasoning**:
Interpret the calculated average evaluation metrics to understand the performance of the recommendation engine and discuss its strengths and weaknesses based on these scores.



In [238]:
print("Interpretation of Evaluation Results:")
print(f"Average Precision@{k_recommendations}: {average_precision:.4f}")
print(f"Average Recall@{k_recommendations}: {average_recall:.4f}")
print(f"Average NDCG@{k_recommendations}: {average_ndcg:.4f}")

print("\nAnalysis:")
print(f"- Precision@{k_recommendations}: On average, {average_precision:.1%} of the top {k_recommendations} recommended movies were considered relevant based on the defined similarity threshold.")
print(f"- Recall@{k_recommendations}: The engine was able to recommend, on average, {average_recall:.1%} of the actual relevant movies in the training set within the top {k_recommendations} recommendations.")
print(f"- NDCG@{k_recommendations}: The ranking of the top {k_recommendations} recommendations aligns reasonably well with the relevance of the items, with an average score of {average_ndcg:.4f}. A higher score indicates better ranking of more relevant items at higher positions.")

print("\nStrengths and Weaknesses:")
print("Based on these metrics, the recommendation engine seems to have a moderate ability to recommend relevant movies (Precision@10). It is able to capture a portion of the potentially relevant movies from the training set (Recall@10). The NDCG@10 score suggests that the more relevant items among the recommendations tend to be ranked higher, which is a positive aspect.")
print("\nPotential Weaknesses:")
print(f"The Recall@{k_recommendations} score ({average_recall:.1%}) indicates that a significant portion of potentially relevant movies in the training set are not being included in the top {k_recommendations} recommendations for a given test movie. This could be due to:")
print("- The chosen similarity threshold (0.2) for defining 'actual relevant items' might be too high, making the set of 'actual relevant items' very small and thus artificially inflating Recall for some cases where only a few relevant items exist.")
print("- The cosine similarity measure based on the current feature representation might not fully capture all nuances of movie similarity.")
print("- The 'max_features' parameter in CountVectorizer might be limiting the vocabulary and thus the richness of the movie representations.")

print(f"\nImpact of Similarity Threshold ({similarity_threshold}):")
print(f"The chosen similarity threshold of {similarity_threshold} directly impacts the definition of 'actual relevant items'. A higher threshold means fewer items in the training set are considered relevant to a test movie. This can make it easier to achieve a high Precision if the few truly similar items are ranked highly, but it can also make it harder to achieve a high Recall if many moderately similar (but still relevant) items are excluded from the 'actual relevant' set.")
print("Adjusting this threshold could significantly change the reported metrics and provide a different perspective on the engine's performance.")

Interpretation of Evaluation Results:
Average Precision@10: 0.7507
Average Recall@10: 0.5877
Average NDCG@10: 0.9647

Analysis:
- Precision@10: On average, 75.1% of the top 10 recommended movies were considered relevant based on the defined similarity threshold.
- Recall@10: The engine was able to recommend, on average, 58.8% of the actual relevant movies in the training set within the top 10 recommendations.
- NDCG@10: The ranking of the top 10 recommendations aligns reasonably well with the relevance of the items, with an average score of 0.9647. A higher score indicates better ranking of more relevant items at higher positions.

Strengths and Weaknesses:
Based on these metrics, the recommendation engine seems to have a moderate ability to recommend relevant movies (Precision@10). It is able to capture a portion of the potentially relevant movies from the training set (Recall@10). The NDCG@10 score suggests that the more relevant items among the recommendations tend to be ranked high

## Summary:

### Data Analysis Key Findings

*   The dataset was split into training (80%) and testing (20%) sets for model evaluation.
*   Functions for calculating Precision@k, Recall@k, and NDCG@k were successfully implemented.
*   The recommendation engine was evaluated on the test set against the training set, using a similarity threshold of 0.2 to define actual relevant items.
*   The average evaluation metrics at k=10 were:
    *   Average Precision@10: 0.7507
    *   Average Recall@10: 0.5877
    *   Average NDCG@10: 0.9647

### Insights or Next Steps

*   While the engine shows good precision and ranking (high Precision and NDCG), the Recall score indicates that it is missing a notable portion of potentially relevant items. Further investigation into the feature representation or the similarity threshold might be beneficial.
*   Consider experimenting with different similarity thresholds for defining "actual relevant items" to understand its impact on the evaluation metrics and gain a more comprehensive view of the engine's performance.


## Summary:

### Data Analysis Key Findings

* The dataset was split into training (80%) and testing (20%) sets for model evaluation.
* Functions for calculating Precision@k, Recall@k, and NDCG@k were successfully implemented.
* The recommendation engine was evaluated on the test set against the training set, using a similarity threshold of 0.2 to define actual relevant items.
* By including more features (production companies, countries, spoken languages) and increasing the `max_features` in the CountVectorizer, the model's performance changed significantly.
* The average evaluation metrics at k=10 for the improved model were:
  * Average Precision@10: {average_precision_improved:.4f} (Increased significantly)
  * Average Recall@10: {average_recall_improved:.4f} (Decreased)
  * Average NDCG@10: {average_ndcg_improved:.4f} (Slightly increased)

### Insights or Next Steps

* The improved model shows very high precision and excellent ranking (high Precision and NDCG), indicating that when it recommends a movie, it's highly likely to be relevant, and the more relevant items are ranked higher.
* However, the decrease in Recall suggests a trade-off: the engine is now recommending a smaller proportion of the actual relevant movies available in the training set. This might mean it's becoming more specialized and less likely to surface a wider variety of relevant content.
* **Further Experimentation:** As suggested in the analysis, experimenting with different similarity thresholds and exploring other vectorization techniques like TF-IDF could help understand and potentially balance the trade-off between Precision and Recall.
* **Qualitative Analysis:** Examining specific examples of recommendations (false positives and false negatives) could provide valuable qualitative insights into why certain recommendations are missed or irrelevant.

This concludes the task of evaluating and improving the recommendation engine based on the initial approach and evaluation metrics.

## Analyze Improved Model Results

Interpret the evaluation results of the improved model to understand the impact of adding more features and increasing `max_features` on the recommendation engine's performance.

In [241]:
print("Interpretation of Improved Model Evaluation Results:")
print(f"Average Precision@{k_recommendations_improved}: {average_precision_improved:.4f}")
print(f"Average Recall@{k_recommendations_improved}: {average_recall_improved:.4f}")
print(f"Average NDCG@{k_recommendations_improved}: {average_ndcg_improved:.4f}")

print("\nAnalysis of Improvements:")
print(f"Compared to the previous evaluation (Precision@{k_recommendations:.4f}, Recall@{k_recommendations:.4f}, NDCG@{k_recommendations:.4f}):")

print(f"- Precision@{k_recommendations_improved}: The average precision has significantly increased to {average_precision_improved:.1%}. This suggests that a much higher percentage of the top {k_recommendations_improved} recommendations are now considered relevant.")
print(f"- Recall@{k_recommendations_improved}: The average recall has decreased to {average_recall_improved:.1%}. This indicates that while the precision of the recommendations has improved, the engine is recommending a smaller proportion of the actual relevant movies in the training set.")
print(f"- NDCG@{k_recommendations_improved}: The average NDCG has slightly increased to {average_ndcg_improved:.4f}. This suggests a slightly better ranking of the relevant items within the top recommendations.")

print("\nDiscussion:")
print("The improvement in Precision and NDCG suggests that by including more features and increasing the `max_features` in the vectorizer, the engine is becoming better at identifying highly relevant movies. The decrease in Recall, however, might imply that the engine is becoming more focused on a narrower set of highly similar items, potentially missing some other relevant but less similar movies.")
print("\nPossible reasons for the observed changes:")
print("- **Increased feature set:** Including production companies, countries, and spoken languages adds more specific information, which can help in identifying more precise matches, leading to higher precision.")
print("- **Increased `max_features`:** A larger vocabulary size in the CountVectorizer allows for a more detailed representation of each movie's content, potentially leading to better distinction between movies and thus higher precision for the top recommendations.")
print("- **Similarity Threshold:** The fixed similarity threshold of 0.2 might be interacting differently with the new, richer feature space. It's possible that a threshold of 0.2 now corresponds to a higher level of similarity than before, thus reducing the size of the 'actual relevant items' set and impacting recall.")

print("\nNext Steps for Further Improvement:")
print("- **Adjust the similarity threshold:** Experiment with different similarity thresholds to see how they impact the trade-off between Precision and Recall.")
print("- **Explore other vectorization techniques:** Try TF-IDF to see if weighting terms differently affects the results.")
print("- **Analyze false positives and false negatives:** Look at examples of recommendations that were not considered relevant (false positives) and actual relevant items that were not recommended (false negatives) to gain more qualitative insights into the engine's behavior.")

Interpretation of Improved Model Evaluation Results:
Average Precision@10: 0.9711
Average Recall@10: 0.1905
Average NDCG@10: 0.9990

Analysis of Improvements:
Compared to the previous evaluation (Precision@10.0000, Recall@10.0000, NDCG@10.0000):
- Precision@10: The average precision has significantly increased to 97.1%. This suggests that a much higher percentage of the top 10 recommendations are now considered relevant.
- Recall@10: The average recall has decreased to 19.0%. This indicates that while the precision of the recommendations has improved, the engine is recommending a smaller proportion of the actual relevant movies in the training set.
- NDCG@10: The average NDCG has slightly increased to 0.9990. This suggests a slightly better ranking of the relevant items within the top recommendations.

Discussion:
The improvement in Precision and NDCG suggests that by including more features and increasing the `max_features` in the vectorizer, the engine is becoming better at identifyi

## Improving the Recommendation Engine

Let's try to improve the recommendation engine by including more features and adjusting the vectorization.

In [240]:
# Re-load the original dataframes to ensure all columns are available
movies_df = pd.read_csv('tmdb_5000_movies.csv')
credits_df = pd.read_csv('tmdb_5000_credits.csv')

# Merge the dataframes
movies = movies_df.merge(credits_df , on ="title")

# Drop rows with missing overview
movies.dropna(subset=['overview'], inplace=True)


# Recreate the 'tags' column with more features
movies['tags'] = movies['overview'].apply(lambda x: x.split()) + \
                 movies['genres'].apply(convert) + \
                 movies['keywords'].apply(convert) + \
                 movies['cast'].apply(convert2) + \
                 movies['crew'].apply(fetch_director) + \
                 movies['production_companies'].apply(convert) + \
                 movies['production_countries'].apply(convert) + \
                 movies['spoken_languages'].apply(convert)


# Create a new DataFrame with relevant columns
new_df_improved = movies[['id', 'title', 'tags']].copy()

# Apply stemming and lowercasing to the new tags
new_df_improved['tags'] = new_df_improved['tags'].apply(lambda x: " ".join(x))
new_df_improved['tags'] = new_df_improved['tags'].apply(lambda x: x.lower())
new_df_improved['tags'] = new_df_improved['tags'].apply(stem)


# Split the improved data into training and testing sets
new_df_train_improved, new_df_test_improved = train_test_split(new_df_improved, test_size=0.2, random_state=42)

# Re-vectorize the data with an increased max_features
cv_improved = CountVectorizer(max_features=10000, stop_words='english') # Increased max_features
vectors_train_improved = cv_improved.fit_transform(new_df_train_improved["tags"]).toarray()
vectors_test_improved = cv_improved.transform(new_df_test_improved["tags"]).toarray()

# Recalculate similarity between test set vectors and training set vectors
similarity_test_train_improved = cosine_similarity(vectors_test_improved, vectors_train_improved)

# Evaluate the improved model
k_recommendations_improved = 10
similarity_threshold_improved = 0.2 # Keep the same threshold for comparison

all_precision_scores_improved = []
all_recall_scores_improved = []
all_ndcg_scores_improved = []

# Iterate through each movie in the test set
for index, row in new_df_test_improved.iterrows():
    test_movie_title = row['title']

    # Get recommended items from the training set using the improved similarity
    try:
        # Find the index of the movie in the test set
        test_movie_index_in_test_df = new_df_test_improved[new_df_test_improved["title"] == test_movie_title].index[0]
        # Get the corresponding row in the similarity_test_train_improved matrix
        distances_to_train = similarity_test_train_improved[new_df_test_improved.index.get_loc(test_movie_index_in_test_df)]

        # Get the indices of the top k most similar movies in the training set
        movies_list = sorted(list(enumerate(distances_to_train)), reverse=True, key=lambda x: x[1])[:k_recommendations_improved]

        # Filter out the movie itself if it's in the training set and among top k
        recommended_items = []
        for i, score in movies_list:
             recommended_movie_title = new_df_train_improved.iloc[i].title
             if recommended_movie_title != test_movie_title:
                 recommended_items.append(recommended_movie_title)
             if len(recommended_items) == k_recommendations_improved:
                 break


        # Determine actual relevant items from the training set based on similarity threshold
        relevant_indices_in_train = [i for i, score in enumerate(distances_to_train) if score >= similarity_threshold_improved]
        actual_relevant_items = set([new_df_train_improved.iloc[i].title for i in relevant_indices_in_train])

        # Calculate metrics
        precision = calculate_precision(actual_relevant_items, recommended_items)
        recall = calculate_recall(actual_relevant_items, recommended_items)
        ndcg = calculate_ndcg(actual_relevant_items, recommended_items)

        all_precision_scores_improved.append(precision)
        all_recall_scores_improved.append(recall)
        all_ndcg_scores_improved.append(ndcg)

    except IndexError:
        print(f"Movie '{test_movie_title}' not found in the test dataset.")


# Calculate average metrics for the improved model
average_precision_improved = sum(all_precision_scores_improved) / len(all_precision_scores_improved) if all_precision_scores_improved else 0
average_recall_improved = sum(all_recall_scores_improved) / len(all_recall_scores_improved) if all_recall_scores_improved else 0
average_ndcg_improved = sum(all_ndcg_scores_improved) / len(all_ndcg_scores_improved) if all_ndcg_scores_improved else 0

print("\nEvaluation Results for Improved Model:")
print(f"Average Precision@{k_recommendations_improved}: {average_precision_improved}")
print(f"Average Recall@{k_recommendations_improved}: {average_recall_improved}")
print(f"Average NDCG@{k_recommendations_improved}: {average_ndcg_improved}")


Evaluation Results for Improved Model:
Average Precision@10: 0.9711018711018712
Average Recall@10: 0.1904722154496127
Average NDCG@10: 0.998960498960499
